prov/gni/src/gnix_eq.c

```diff
@@ -78,20 +79,181 @@ static int gnix_verify_eq_attr(struct fi_eq_attr *attr)
  */
 static void free_queue(struct slist *list)
 {
-	slist_entry *current = NULL;
-	slist_entry *next = NULL;
+	struct slist_entry *current = NULL;
+	struct slist_entry *next = NULL;
 	struct gnix_event *temp = NULL;
 
 	if (slist_empty(list))
 		return;
 
-	for (current = list.head; current; current = next) {
+	for (current = list->head; current; current = next) {
 		next = current->next;
 		temp = container_of(current, struct gnix_event, entry);
+
+		/*
+		 * For the event queue this is the extraneous data that was
+		 * copied in. For the error queue this is the actual
+		 * fi_eq_err_entry.
+		 */
+		free(temp->buf);
 		free(temp);
 	}
 }
 
+static ssize_t queue_read(struct fid_eq *eq, uint32_t *event, void *buf,
+			  size_t len, uint64_t flags, enum q_type type)
+{
+	struct gnix_fid_eq *gnix_eq = NULL;
+	struct gnix_event *ev = NULL;
+	struct slist *queue;
+	int ret = len;
+
+	gnix_eq = container_of(eq, struct gnix_fid_eq, eq_fid);
+	if (!gnix_eq) {
+		ret = -FI_EINVAL;
+		goto err;
+	}
+
+	fastlock_acquire(&gnix_eq->lock);
+
+	switch (type) {
+	case EVENT:
+		if (!slist_empty(&gnix_eq->err_queue)) {
+			ret = -FI_EAVAIL;
+			goto cleanup;
+		}
+
+		if (slist_empty(&gnix_eq->ev_queue)) {
+			ret = -FI_EAGAIN;
+			goto cleanup;
+		}
+
+		queue = &gnix_eq->ev_queue;
+		break;
+	case ERROR:
+		if (slist_empty(&gnix_eq->err_queue)) {
+			ret = -FI_EAGAIN;
+			goto cleanup;
+		}
+
+		queue = &gnix_eq->err_queue;
+		break;
+	default:
+		ret = -FI_EINVAL;
+		goto cleanup;
+	}
+
+	ev = container_of(queue->head, struct gnix_event, entry);
+	if (!event) {
+		ret = -FI_EINVAL;
+		goto cleanup;
+	}
+
+	if (len < ev->len) {
+		ret = -FI_ETOOSMALL;
+		goto cleanup;
+	}
+
+	if (type == EVENT)
+		*event = ev->type;
+
+	memcpy(buf, ev->buf, len);
+
+	/*
+	 * Remove if FI_PEEK isn't specified.
+	 */
+	if (!(flags & FI_PEEK)) {
+		ev = container_of(slist_remove_head(&gnix_eq->ev_queue),
+				  struct gnix_event, entry);
+		free(ev->buf);
+		free(ev);
+	}
+
+	fastlock_release(&gnix_eq->lock);
+
+	return ret;
+
+cleanup:
+	fastlock_release(&gnix_eq->lock);
+err:
+	return ret;
+}
+
+static ssize_t queue_write(struct fid_eq *eq, uint32_t event, void *buf,
+			   size_t len, uint64_t flags, enum q_type type)
+{
+	struct gnix_fid_eq *gnix_eq = NULL;
+	struct gnix_event *q_entry = NULL;
+	struct slist *queue = NULL;
+	ssize_t ret = len;
+
+	q_entry = calloc(1, sizeof(*q_entry));
+	if (!q_entry) {
+		ret = -FI_ENOMEM;
+		goto err;
+	}
+
+	gnix_eq = container_of(eq, struct gnix_fid_eq, eq_fid);
+	if (!gnix_eq) {
+		ret = -FI_EINVAL;
+		goto cleanup_q_entry;
+	}
+
+	switch (type) {
+	case EVENT:
+		queue = &gnix_eq->ev_queue;
+		break;
+	case ERROR:
+		queue = &gnix_eq->err_queue;
+		break;
+	default:
+		ret = -FI_EINVAL;
+		goto cleanup_q_entry;
+	}
+
+	q_entry->len = len;
+	q_entry->buf = buf;
+	q_entry->type = event;
+	q_entry->flags = flags;
+
+	fastlock_acquire(&gnix_eq->lock);
```

Comment:
I'm not sure how to handle the errrors for fastlock_init, fastlock_acquire, and fastlock_relese. They return regular errno codes and there isn't a conversion function for converting to fabric codes. Should I just log the output of strerror and return -FI_EBUSY? I checked around the project and the others that use fastlock* don't check the return status. I'm not sure if it's important to be checking or not...

@hppritcha @sungeunchoi